reports.py

```diff
@@ -74,6 +74,104 @@ def average_rate_per_genre(context: dict[str, DataFrame]) -> DataFrame:
 
     return df
 
+def most_popular_title_of_each_actor(context: dict[str, DataFrame]) -> DataFrame:
+    principals = context["principals"]
+    ratings = context["ratings"]
+    name_basics = context["name_basics"]
+    title_basics = context["title_basics"]
+
+    df = (
+        principals.alias("p")
+        .filter(
+            (f.col("p.category") == "actor")
+            | (f.col("p.category") == "actress")
+        )
+        .join(ratings.alias("r"), f.col("p.tconst") == f.col("r.tconst"))
+        .join(name_basics.alias("nb"), f.col("p.nconst") == f.col("nb.nconst"))
+        .join(title_basics.alias("tb"), f.col("p.tconst") == f.col("tb.tconst"))
+        .groupBy("nb.primaryName", "tb.primaryTitle")
+        .agg(
+            f.max("r.numVotes").alias("NumVotes"),
+        )
+        .select(
+            f.col("nb.primaryName").alias("Name"),
+            f.col("tb.primaryTitle").alias("Title"),
+            f.col("NumVotes"),
+        )
+        .orderBy(f.col("NumVotes").desc())
+    )
+
+    return df
+
+def last_film_of_each_director(context: dict[str, DataFrame]) -> DataFrame:
+    title_basics = context["title_basics"]
+    principals = context["principals"]
+    name_basics = context["name_basics"]
+    window = Window.partitionBy("primaryName").orderBy(f.col("startYear").desc())
+
+    df = (
+        title_basics.alias("tb")
+        .filter(f.col("tb.startYear") < 2024)
+        .join(principals.alias("p"), f.col("tb.tconst") == f.col("p.tconst"))
+        .join(name_basics.alias("nb"), f.col("nb.nconst") == f.col("p.nconst"))
+        .filter(f.col("p.category") == "director")
+        .withColumn("rowNumber", f.row_number().over(window))
+        .filter(f.col("rowNumber") == 1)
+        .drop(f.col("rowNumber"))
+        .select(
+            f.col("nb.primaryName").alias("DirectorName"),
+            f.col("tb.startYear").alias("FilmYear"),
+            f.col("tb.primaryTitle").alias("Title"),
+        )
+        .orderBy(f.col("FilmYear").desc())
+    )
+
+    return df
+
+
+def actors_who_are_younger_thirty(context: dict[str, DataFrame]) -> DataFrame:
+    name_basics = context["name_basics"]
+
+    name_basics_transformed = (
+        name_basics
+        .withColumn("profession", f.split("primaryProfession", ","))
+        .drop("primaryProfession")
+        .withColumn("profession", f.explode("profession"))
```

Comment:
you can simplify this into one row
`.withColumn("profession",f.explode(f.split("primaryProfession", ",")))`